src/Utils.hs

```diff
@@ -145,3 +204,125 @@ pconvertChecked x = ptryFrom x fst
+
+psignedByOwner :: Term s (PScriptContext :--> PAddress :--> PBool)
+psignedByOwner = plam $ \ctx owner ->
+  pmatch (pfield @"credential" # owner) $ \case
+    PPubKeyCredential ((pfield @"_0" #) -> pkh) ->
+      pelem @PBuiltinList # pkh # (pfield @"signatories" # (pfield @"txInfo" # ctx))
+    _ ->
+      pcon PFalse
+
+pvalueHasChangedByLovelaces :: Term s (PValue 'Sorted 'Positive :--> PValue 'Sorted 'Positive :--> PInteger :--> PBool)
+pvalueHasChangedByLovelaces = plam $ \inVal outVal change ->
+  pif
+    (change #== 0)
+    (outVal #== inVal)
+    (pforgetPositive outVal #== (pforgetPositive inVal <> (psingleton # padaSymbol # padaToken # change)))
+
+presolveMapToList ::
+  forall
+    (anyOrder :: KeyGuarantees)
+    (a :: PType)
+    (b :: PType)
+    (s :: S).
+  Term s (PMap anyOrder a b) ->
+  Term s (PBuiltinList (PBuiltinPair (PAsData a) (PAsData b)))
+presolveMapToList m = pmatch m $ \(PMap l) -> l
+
+{- | Converts a `PValue` to a `PBuiltinList`. Does not convert the inner `PMap`
+of token names and quantities to a list.
+-}
+presolveValueToList ::
+  forall
+    (anyOrder :: KeyGuarantees)
+    (anyAmount :: AmountGuarantees)
+    (s :: S).
+  Term s (PValue anyOrder anyAmount) ->
+  Term s (PBuiltinList (PBuiltinPair (PAsData PCurrencySymbol) (PAsData (PMap anyOrder PTokenName PInteger))))
+presolveValueToList v =
+  pmatch v $ \(PValue v') ->
+    pmatch (presolveMapToList v') $ \kvs -> pcon kvs
+
+{- | Get the head of the list if the list contains exactly one element,
+otherwise error.
+-}
+pheadSingleton ::
+  (PListLike list, PElemConstraint list a) =>
+  Term s (list a) ->
+  Term s a
+pheadSingleton =
+  pelimList
+    (pelimList (\_ _ -> ptraceError "List contains more than one element."))
+    (ptraceError "List is empty.")
+
+-- | Helper function for converting intermediary datatypes.
+psingleAssetToTriple ::
+  forall
+    (anyOrder :: KeyGuarantees)
+    (s :: S).
+  Term
+    s
+    ( PBuiltinPair
+        (PAsData PCurrencySymbol)
+        (PAsData (PMap anyOrder PTokenName PInteger))
+        :--> PTriple PCurrencySymbol PTokenName PInteger
+    )
+psingleAssetToTriple = plam $ \asset ->
+  let
+    cs = pfstBuiltin # asset
+    tnQtyMap = psndBuiltin # asset
+    tnQtyPairs = presolveMapToList $ pfromData tnQtyMap
+   in
+    plet (pheadSingleton tnQtyPairs) $ \tnQtyPair ->
+      let
+        tn = pfstBuiltin # tnQtyPair
+        qty = psndBuiltin # tnQtyPair
+       in
+        pcon (PTriple (pfromData cs) (pfromData tn) (pfromData qty))
+
+-- | Grabs the singular asset in a given `PValue`, ignoring its ADA.
+pgetSingleAssetApartFromADA ::
+  forall
+    (anyOrder :: KeyGuarantees)
+    (anyAmount :: AmountGuarantees)
+    (s :: S).
+  Term s (PValue anyOrder anyAmount) ->
+  Term s (PTriple PCurrencySymbol PTokenName PInteger)
+pgetSingleAssetApartFromADA v =
+  -- not using `pelimList` as it would've lead to evaluation of the head
+  psingleAssetToTriple #$ pheadSingleton $ ptail # presolveValueToList v
```

Comment:
This only works for `'Sorted` key guarantee!

Reply:
Very nice catch! Thank you :pray: